llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp

```diff
@@ -1098,6 +1094,38 @@ void AsmPrinter::emitBBAddrMapSection(const MachineFunction &MF) {
   // Emit the total number of basic blocks in this function.
   OutStreamer->emitULEB128IntValue(MF.size());
   const Function &F = MF.getFunction();
+
+  // During code-gen, LLVM may merge more than one BasicBlock into a single
```

Comment:
I think this is a really critical comment, so I want to make sure I've understood it. I think it can be simplified to something like:

```
LLVM's codegen can can merge multiple (IR-level) `BasicBlock`s into a single
(machine code) `MachineBasicBlock`. Unfortunately,
`MachineBasicBlock::getBasicBlock()` only returns the first `BasicBlock` in the
merged sequence, so we have to find the remaining `BasicBlock`(s) (if any!)
in the merged sequence another way. We do so in two steps:

1. We create a set `MergedBBs` which is the set of `MachineBasicBlock`s that are *not*
returned by `MachineBasicBlock::getBasicBlock()` -- in other words, it's the set of
`BasicBlock`s that have been merged into a predecessor during codegen.
2. For each `BasicBlock` *x* returned by `MachineBasicBlock::getBasicBlock` we check
if it is terminated by an unconditional branch. If that unconditional branch points to a
block *y* in `MergedBBs`, we know that *x* and *y* were merged during codegen.
[Note that we repeat this step again for *y* until we have reached a `BasicBlock`
without an unconditional branch to a block in `MergedBBs`.]
```

Reply:
> which is the set of `BasicBlock`s ...
Apart from that t his sounds fine to me.

Reply:
Aside from @ptersilie's comment:

 - repetition of `in the merged sequence` can be removed.
 - `MergedBBs which is the set of MachineBasicBlocks` -- It's a set of IR `BasicBlocks` (`std::set<const BasicBlock *> MergedBBs`).
 - `*x*` the LLVM convention for blocks, is to use a `BB` or `MBB` suffix.

Apart from those the comment is equivalent (it just sounds like Laurie wrote it in a more conversational style :P).

Reply:
Also, I think we should be careful with the term IR. The codegen's IR is still an IR.

Reply:
I admit that "aforementioned" was too much for me :p It's up to you which version you consider is clearer to readers -- I'm happy either way.

Reply:
> Also, I think we should be careful with the term IR. The codegen's IR is still an IR.

Can we find some way of differentiating the two?

Reply:
We can just drop the bits in brackets. The data-types make it clear. `Machine*` is code-gen.